### PR TITLE
Optimize BN254 multi-pair pairing checks

### PR DESCRIPTION
**Pull request by kamilchodola**
## Changes

- Optimize the BN254 (`alt_bn128`) pairing-check precompile (EIP-197, address `0x08`) by dispatching on pair count:
  - **1 pair** → a dedicated single-pairing fast path.
  - **2–32 pairs** → a vectorized multi-Miller-loop (`mclBn_millerLoopVec`) with a single final exponentiation, using a `[SkipLocalsInit]` stack buffer (capped at 32 pairs to bound stack usage).
  - **>32 pairs** → the original scalar accumulation, so attacker-controlled large inputs never grow the stack allocation.
- Hoist the point-at-infinity (`isZero`) check out of `DeserializeG1`/`DeserializeG2` so the 64/128-byte input scan is reused instead of re-scanning the larger deserialized struct.
- Add unit tests covering the new pair-count dispatch boundaries (see Testing).

All three paths compute the same `finalExp(∏ᵢ e(g1ᵢ, g2ᵢ))`; point validation/subgroup checks (`mclBnG2_isValidOrder`) and gas costs are unchanged.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

The existing `BN254PairingCheckPrecompileTests` vectors only reached 8 pairs, leaving the 32-pair vector boundary and the `>32` scalar fallback unexercised. Added `Multi_pair_boundary` cases that cross the boundary using inputs whose pairing product is known analytically — a repeated EIP-197 two-point match (`e(P,Q)·e(-P,Q)=1`) plus points at infinity — so the vector and scalar paths are both validated against ground truth on `true`, not-one, and all-infinity results:

| Pairs | Path | Construction | Expected |
|---|---|---|---|
| 32 | vectorized | 16× two-point match | `1` |
| 33 | scalar | 16× match + 1 infinity (skipped) | `1` |
| 34 | scalar | 17× match | `1` |
| 33 | scalar | 16× match + 1 pair with product ≠ 1 | `0` |
| 33 | scalar | 33× infinity (all skipped) | `1` |

All pass (`5/5`); full fixture `50/50`.

### Benchmark data

Gas-benchmarks repricing workflow, `repricings_compute/jochemnet`, fork Amsterdam, filter `test_ec_pairing[...num_pairs_12]`, identical Docker configuration for both runs (single run, restart between tests, release `amsterdam-repricings-v5.2.0`):

- **This branch** `nethermindeth/nethermind:perf-ec-pairing-amsterdam` (`1.39.0-unstable+c2fc958b`) — [run 27009614135](https://github.com/NethermindEth/gas-benchmarks/actions/runs/27009614135)
- **master** `nethermindeth/nethermind:master` (`1.39.0-unstable+69aeab7e`, the immediate parent) — [run 27010410713](https://github.com/NethermindEth/gas-benchmarks/actions/runs/27010410713)

`engine_newPayloadV5` average (block processing time), lower is better:

| Gas limit | master | this branch | Δ |
|---:|---:|---:|---:|
| 100M | 253.2 ms | 192.7 ms | **−23.9%** |
| 120M | 287.4 ms | 224.0 ms | −22.1% |
| 140M | 283.7 ms | 242.2 ms | −14.6% |
| 160M | 310.0 ms | 253.6 ms | −18.2% |
| 180M | 313.3 ms | 264.3 ms | −15.6% |
| 200M | 319.0 ms | 280.7 ms | −12.0% |
| 220M | 338.3 ms | 283.0 ms | −16.4% |
| 240M | 385.9 ms | 295.8 ms | −23.3% |
| 260M | 405.1 ms | 323.6 ms | −20.1% |
| 280M | 435.5 ms | 368.7 ms | −15.3% |
| 300M | 496.3 ms | 423.5 ms | −14.7% |
| **avg** | **348.0 ms** | **286.6 ms** | **−17.6%** |

Every gas point improves (12–24%, mean −17.8%). Both runs are single-shot (`runs=1`), so individual points carry run-to-run noise, but the consistent monotonic improvement across all 11 points reflects a real speedup rather than variance.

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs [hunk 1]
```diff
@@ -87,68 +88,145 @@ internal static bool CheckPairing(byte[] output, ReadOnlySpan<byte> input)
...
+    private static bool CheckPairingVector(byte[] output, byte* data, int pairCount)
     {
-        ref byte start = ref Unsafe.As<T, byte>(ref Unsafe.AsRef(in data));
-        ReadOnlySpan<byte> span = MemoryMarshal.CreateReadOnlySpan(in start, sizeof(T));
-        return span.IndexOfAnyExcept((byte)0) < 0;
+        const int UlongSize = sizeof(ulong);
+        int g1UlongCount = sizeof(mclBnG1) / UlongSize;
+        int g2UlongCount = sizeof(mclBnG2) / UlongSize;
+        ulong* g1Buffer = stackalloc ulong[pairCount * g1UlongCount];
+        ulong* g2Buffer = stackalloc ulong[pairCount * g2UlongCount];
```
[the remark this answers is not in the document]
Comment (kamilchodola) [reply src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs]: Addressed in 668b9cfe33: `CheckPairingVector` now allocates the scratch buffers directly in bytes (`stackalloc byte[pairCount * sizeof(mclBnG1)]` / `mclBnG2`), so the allocation size always matches the `sizeof` write stride and no longer relies on the struct size being a multiple of 8.

File: src/Nethermind/Nethermind.Evm.Test/BN254PairingCheckPrecompileTests.cs [hunk 2]
```diff
@@ -53,4 +55,53 @@ public class BN254PairingCheckPrecompileTests : PrecompileTests<BN254PairingChec
...
+            builder.Append(pair);
+        return builder.ToString();
+    }
+
+    private static string InfinityPairs(int count) => new('0', PairHexLength * count);
+
+    private static IEnumerable<TestCaseData> MultiPairBoundaryCases()
+    {
+        // 32 pairs == MaxStackPairCount: vectorized path; 16 cancelling blocks multiply to one.
+        yield return new TestCaseData(Repeat(TwoPairsProductOne, 16), ResultOne, true)
```
[the remark this answers is not in the document]
Comment (kamilchodola) [reply src/Nethermind/Nethermind.Evm.Test/BN254PairingCheckPrecompileTests.cs:93]: Addressed in 668b9cfe33: added an explicit `CheckPairing_4_pairs_product_not_one_vector_path` case (`TwoPairsProductOne + OnePairNotOne + InfinityPairs(1)`, 4 pairs, expecting `ResultNotOne`) so the vectorized path is guarded against short-circuiting to an always-one result.

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs [hunk 3]
```diff
@@ -87,68 +88,141 @@ internal static bool CheckPairing(byte[] output, ReadOnlySpan<byte> input)
...
+            if (pairCount <= MaxStackPairCount)
             {
-                if (!DeserializeG1(data + i, out mclBnG1 g1))
-                    return false;
+                return CheckPairingVector(output, data, pairCount);
+            }
 
-                if (!DeserializeG2(data + i + 64, out mclBnG2 g2))
-                    return false;
+            return CheckPairingScalar(output, data, pairCount);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs]: 1. do a return switch expression!
2. If pairCount > MaxStackPairCount, can we do Vector of first 32 pairs (N times if it is more than 64 ect), and then do scalar on the reminder?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs [hunk 4]
```diff
@@ -87,68 +88,121 @@ internal static bool CheckPairing(byte[] output, ReadOnlySpan<byte> input)
...
+    {
+        // Process the pairs in chunks of at most MaxStackPairCount so the scratch buffers stay a fixed,
+        // input-independent size on the stack (the >MaxStackPairCount case never grows the allocation),
+        // while still feeding the vectorized multi-Miller-loop for every pair rather than falling back to
+        // a per-pair scalar loop. Each chunk's Miller-loop product is multiplied into a running GT
+        // accumulator and a single final exponentiation is applied at the end — finalExp(∏ ML) is invariant
+        // to how the product is batched.
+        // Allocate in bytes so the buffer size matches the write stride (sizeof) exactly, regardless of struct padding.
+        int chunkCapacity = Math.Min(pairCount, MaxStackPairCount);
+        byte* g1Bytes = stackalloc byte[chunkCapacity * sizeof(mclBnG1)];
```
Comment (Marchhill) [inline src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs:135]: not sure it's safe to stackalloc here?

**Review by Marchhill — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs (hunk 3 above)
Comment (kamilchodola) [reply src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs]: Both addressed in `b234e17047`:

1. **Switch expression** — the pair-count dispatch is now a `return` switch (`1 => CheckPairingSingle`, `_ => CheckPairingVector`).

2. **Vector the big path instead of scalar fallback** — done, with one deviation from the literal suggestion: rather than "vector the first 32 then scalar the remainder", `CheckPairingVector` now processes the pairs in chunks of `MaxStackPairCount`, runs **every** chunk (including the remainder) through `mclBn_millerLoopVec`, and combines the per-chunk Miller-loop products in GT before a single `finalExp`. Vectorizing the remainder too is strictly faster than scalar for it, keeps the scratch buffer a fixed `min(pairCount, 32)` size on the stack regardless of input, and lets me delete `CheckPairingScalar` entirely (the `>32` path is no longer a separate code path). `finalExp(∏ ML)` is invariant to how the product is batched, so the result is unchanged.

Tests updated to exercise multi-chunk accumulation (the `>32` cases now validate cross-chunk GT combination, including a cross-chunk not-one result); all 90 BN254 tests pass.

Perf note: I re-ran `repricings_compute/jochemnet` `test_ec_pairing num_pairs_12` to check for regressions. `num_pairs_12` is a single chunk (≤32), so it is behaviorally identical to the previous vector path — a contemporaneous A/B (this commit vs its parent `668b9cfe33`, both built/run together) came out at −0.8% AVG, within run-to-run noise. The chunked change only alters the `>32`-pair path, which the standard fixture set does not reach, so it is a worst-case/bounded-stack improvement rather than a `num_pairs_12` speedup.

File: src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs (hunk 4 above)
Comment (Marchhill) [reply src/Nethermind/Nethermind.Evm.Precompiles/BN254.cs:135]: actually looks ok

**Review by Marchhill — APPROVED** [review]
